### PR TITLE
[tests] Add `Aspire.Playground.Tests`

### PR DESCRIPTION
**Pull request by radical**
Add wrapper tests for apps in `playground/`.

- The tests are based on https://github.com/dotnet/aspire-samples/blob/aee52a7a08ca3433bd1e54e8a90a7febdc852ebd/tests/SamplesIntegrationTests/AppHostTests.cs .
- And test infrastructure from https://github.com/dotnet/aspire-samples/tree/aee52a7a08ca3433bd1e54e8a90a7febdc852ebd//tests/SamplesIntegrationTests/Infrastructure has been copied to the project here

This effectively increases coverage for various components like Qdrant, and Seq, and their hosting bits.

Details:

- Skip dashboard project reference to playground apps when running building for tests
- Use `MapDefaultEndpoints` to add `/alive` and `/health` endpoints needed for testing
- Rename `playground/TestShop/ServiceDefaults` to `playground/TestShop/TestShop.ServiceDefaults` to disambiguate the project binaries in `artifacts`
- Only a few of the playground apps are being tested here. More will be added in follow up PRs.

Contributes to https://github.com/dotnet/aspire/issues/4297 .
 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/5208)

### CODE REVIEW

**Comment by mitchdenny** [comment]
As part of this it would be good to ensure that the the content of aspire-manifest.json has not changed in each of the playground projects. We want to detect if the apphost is generating a manifest different to what is checked in so that we can identify changes that we didn't intend (it also helps with the review process).

**Comment by radical** [comment]
> As part of this it would be good to ensure that the the content of aspire-manifest.json has not changed in each of the playground projects. We want to detect if the apphost is generating a manifest different to what is checked in so that we can identify changes that we didn't intend (it also helps with the review process).

Added to https://github.com/dotnet/aspire/issues/5210

File: tests/Aspire.Playground.Tests/AppHostTests.cs [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+        await app.StopAsync();
+    }
+
+    public static TheoryData<string> AppHostAssemblies()
+    {
+        var appHostAssemblies = GetPlaygroundAppHostAssemblyPaths();
+        var theoryData = new TheoryData<string>();
+        foreach (var asm in appHostAssemblies)
+        {
+            if (string.IsNullOrEmpty(s_appHostNameFilter) || !asm.Contains(s_appHostNameFilter, StringComparison.OrdinalIgnoreCase))
```
Comment (radical) [inline tests/Aspire.Playground.Tests/AppHostTests.cs:115]:
```suggestion
            if (string.IsNullOrEmpty(s_appHostNameFilter) || asm.Contains(s_appHostNameFilter, StringComparison.OrdinalIgnoreCase))
```

File: playground/TestShop/TestShop.sln [hunk 2]
```diff
@@ -9,7 +9,7 @@ Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "BasketService", "BasketServ
 EndProject
 Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "CatalogService", "CatalogService\CatalogService.csproj", "{F8F2A226-59D7-4023-A7BE-602B6B0366D2}"
 EndProject
-Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "ServiceDefaults", "ServiceDefaults\ServiceDefaults.csproj", "{938E6618-3061-4908-82AB-A2463F6B7BE8}"
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "TestShop.ServiceDefaults", "TestShop.ServiceDefaults\TestShop.ServiceDefaults.csproj", "{938E6618-3061-4908-82AB-A2463F6B7BE8}"
```
Comment (eerhardt) [inline playground/TestShop/TestShop.sln:12]: Is this change necessary? Why do we need to rename this project?

File: playground/Directory.Build.targets [hunk 3]
```diff
@@ -1,11 +1,18 @@
 <Project>
   <Import Project="$([MSBuild]::GetPathOfFileAbove('Directory.Build.targets', '$(MSBuildThisFileDirectory)../'))" />
 
-  <ItemGroup Condition="'$(BuildForTest)' != 'true' and '$(IsAspireHost)' == 'true' and '$(SkipDashboardProjectReference)' != 'true'">
+  <PropertyGroup>
+    <IncludeTestPackages Condition="'$(IncludeTestPackages)' == '' and ('$(IsTestProject)' == 'true' or '$(IsTestSupportProject)' == 'true') and '$(SkipDashboardProjectReference)' == 'true'">true</IncludeTestPackages>
```
Comment (eerhardt) [inline playground/Directory.Build.targets:5]: When is `'$(IsTestProject)' == 'true' or '$(IsTestSupportProject)' == 'true'` true for these playground projects?

Comment (eerhardt) [reply playground/Directory.Build.targets:5]: The `SkipDashboardProjectReference` check here is a little surprising. Why that check?

Comment (eerhardt) [reply playground/Directory.Build.targets:5]: Actually, why do we need to set `IncludeTestPackages` here at all?

File: tests/Aspire.Playground.Tests/README.md [hunk 4]
```diff
@@ -0,0 +1,5 @@
+# Aspire.Playground.Tests
+
+Runs apps in `playground/` and hits some endpoints on them.
+
+Use `TEST_PLAYGROUND_APPHOST_FILTER` to run tests for specific playground apps. For example, `TEST_PLAYGROUND_APPHOST_FILTER=proxy`.
```
Comment (eerhardt) [inline tests/Aspire.Playground.Tests/README.md:5]: what is `proxy`? Can we use a real app that works? Like TestShop?

File: tests/Aspire.Playground.Tests/.runsettings [hunk 5]
```diff
@@ -0,0 +1,23 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (eerhardt) [inline tests/Aspire.Playground.Tests/.runsettings:1]: Do we need this copy of this file? Can we use the default one in the root?

File: tests/helix/send-to-helix-inner.proj [hunk 6]
```diff
@@ -10,6 +10,8 @@
     <DotNetCliPackageType>sdk</DotNetCliPackageType>
     <NeedsCreateDotNetDevScripts Condition="'$(OS)' != 'Windows_NT'">true</NeedsCreateDotNetDevScripts>
     <HasDocker Condition="'$(OS)' != 'Windows_NT'">true</HasDocker>
+    <!-- Disable playwright tests on helix/linux. Issue: https://github.com/dotnet/aspire/issues/4623 -->
```
Comment (eerhardt) [inline tests/helix/send-to-helix-inner.proj:13]: Is this playright stuff related to this change?

File: tests/helix/send-to-helix-buildonhelixtests.targets [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+    <ItemGroup>
+      <_TestRunCommandArguments Remove="@(_TestRunCommandArguments)" />
+
+      <!-- Using `dotnet test` for the project directly here -->
+      <_TestRunCommandArguments Include="dotnet test" />
+      <_TestRunCommandArguments Include="-s .runsettings" />
+      <_TestRunCommandArguments Include="--results-directory $(_HelixLogsPath)" />
+      <_TestRunCommandArguments Include="--blame-hang" />
+      <_TestRunCommandArguments Include="--blame-hang-dump-type" />
+      <_TestRunCommandArguments Include="full" />
```
Comment (eerhardt) [inline tests/helix/send-to-helix-buildonhelixtests.targets:41]:
```suggestion
      <_TestRunCommandArguments Include="--blame-hang-dump-type full" />
```

Would it be more readable if these args were together? SAme for timeout and crash dump below.

File: tests/Aspire.Playground.Tests/Aspire.Playground.Tests.csproj [hunk 8]
```diff
@@ -0,0 +1,52 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>$(NetCurrent)</TargetFramework>
+
+    <IsTestProject>true</IsTestProject>
```
Comment (eerhardt) [inline tests/Aspire.Playground.Tests/Aspire.Playground.Tests.csproj:6]:
```suggestion
```

Why is this needed? It should default to a test project cuz of the name.

File: tests/Aspire.Playground.Tests/Aspire.Playground.Tests.csproj [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+    <TestsSharedDir>$(MSBuildThisFileDirectory)..\Shared\</TestsSharedDir>
+    <RunSettingsFilePath>$(MSBuildThisFileDirectory).runsettings</RunSettingsFilePath>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <Compile Include="$(TestsSharedDir)WorkloadTesting\TestOutputWrapper.cs" />
+    <Compile Include="$(TestsSharedDir)WorkloadTesting\EnvironmentVariables.cs" />
+    <Compile Include="$(TestsSharedDir)Logging\*.cs" />
+
+    <AspireProjectOrPackageReference Include="Aspire.Hosting" />
```
Comment (eerhardt) [inline tests/Aspire.Playground.Tests/Aspire.Playground.Tests.csproj:25]: Do we rebuild the Aspire.Playground.Tests in Helix? If so, why? Why don't we build this project on the build machine, and then run the already built tests on Helix?

File: playground/TestShop/TestShop.sln (hunk 2 above)
Comment (radical) [reply playground/TestShop/TestShop.sln:12]: This gets `artifacts/bin/ServiceDefaults` bindir. This rename disambiguates it. It's not *strictly* required though.

File: tests/Aspire.Playground.Tests/README.md (hunk 4 above)
Comment (radical) [reply tests/Aspire.Playground.Tests/README.md:5]: `ProxylessEndToEnd`. Yep, we can use `TestShop`.

File: tests/Aspire.Playground.Tests/.runsettings (hunk 5 above)
Comment (radical) [reply tests/Aspire.Playground.Tests/.runsettings:1]: This is to allow the source being buildable from outside the repo, as we wouldn't have `eng/testing/.runsettings` available there.

The alternative would be to copy it for the archive, and also have a `<RunSettingsFilePath Condition="'$(RepoRoot)' == ''">$(MSBuildThisFileDirectory).runsettings</RunSettingsFilePath>`.

File: tests/Aspire.Playground.Tests/Aspire.Playground.Tests.csproj [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+
+    <ProjectReference Include="$(PlaygroundSourceDir)ProxylessEndToEnd/ProxylessEndToEnd.AppHost/ProxylessEndToEnd.AppHost.csproj" AdditionalProperties="SkipDashboardProjectReference=true" />
+    <ProjectReference Include="$(PlaygroundSourceDir)Qdrant/Qdrant.AppHost/Qdrant.AppHost.csproj" AdditionalProperties="SkipDashboardProjectReference=true" />
+    <ProjectReference Include="$(PlaygroundSourceDir)TestShop/TestShop.AppHost/TestShop.AppHost.csproj" AdditionalProperties="SkipDashboardProjectReference=true" />
+    <ProjectReference Include="$(PlaygroundSourceDir)mongo/Mongo.AppHost/Mongo.AppHost.csproj" AdditionalProperties="SkipDashboardProjectReference=true" />
+    <ProjectReference Include="$(PlaygroundSourceDir)mysql/MySqlDb.AppHost/MySqlDb.AppHost.csproj" AdditionalProperties="SkipDashboardProjectReference=true" />
+    <ProjectReference Include="$(PlaygroundSourceDir)nats/Nats.AppHost/Nats.AppHost.csproj" AdditionalProperties="SkipDashboardProjectReference=true" />
+    <ProjectReference Include="$(PlaygroundSourceDir)seq/Seq.AppHost/Seq.AppHost.csproj" AdditionalProperties="SkipDashboardProjectReference=true" />
```
[the remark this answers is not in the document]
Comment (radical) [reply tests/Aspire.Playground.Tests/Aspire.Playground.Tests.csproj:37]: Because not all of them work right now. Issues like https://github.com/dotnet/aspire/issues/5219, https://github.com/dotnet/aspire/pull/5218, some of the projects requiring azure creds, some failing with other issues that need to be investigated.

File: playground/TestShop/TestShop.sln (hunk 2 above)
Comment (eerhardt) [reply playground/TestShop/TestShop.sln:12]: Ok, I'm fine with the change, I just wanted to ensure I understood why we were making it.

File: tests/Aspire.Playground.Tests/.runsettings (hunk 5 above)
Comment (eerhardt) [reply tests/Aspire.Playground.Tests/.runsettings:1]: Can we just have the root `.runsettings` file checked into the repo, and then the infrastructure copies it to the correct place in the helix payload?

File: tests/Aspire.Playground.Tests/Aspire.Playground.Tests.csproj [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+
+  <ItemGroup>
+    <Compile Include="$(TestsSharedDir)WorkloadTesting\TestOutputWrapper.cs" />
+    <Compile Include="$(TestsSharedDir)WorkloadTesting\EnvironmentVariables.cs" />
+    <Compile Include="$(TestsSharedDir)Logging\*.cs" />
+
+    <AspireProjectOrPackageReference Include="Aspire.Hosting" />
+    <AspireProjectOrPackageReference Include="Aspire.Hosting.AppHost" />
+    <AspireProjectOrPackageReference Include="Aspire.Hosting.Azure" />
+    <AspireProjectOrPackageReference Include="Aspire.Hosting.NodeJs" />
```
[the remark this answers is not in the document]
Comment (radical) [reply tests/Aspire.Playground.Tests/Aspire.Playground.Tests.csproj:28]: re:nodejs, https://github.com/dotnet/aspire/blob/575b0f8fb04bda4fb60fe6e8b16425934053dafa/tests/Aspire.Playground.Tests/Infrastructure/DistributedApplicationExtensions.cs#L145-L146

.. but we can drop the reference by checking the type name instead of the type itself.

[the remark this answers is not in the document]
Comment (radical) [reply tests/Aspire.Playground.Tests/Aspire.Playground.Tests.csproj:28]: Azure, and AppHost can be dropped 👍

File: tests/Aspire.Playground.Tests/Infrastructure/PasswordGenerator.cs [hunk 12]
```diff
@@ -0,0 +1,145 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+using System.Security.Cryptography;
+
+namespace SamplesIntegrationTests.Infrastructure;
+
+internal static class PasswordGenerator
```
Comment (eerhardt) [inline tests/Aspire.Playground.Tests/Infrastructure/PasswordGenerator.cs:10]: We shouldn't need to copy this. We should be able to reuse what's already in the repo

https://github.com/dotnet/aspire/blob/024b48bf581f6790b0878128c6acff7863d3d5f7/src/Aspire.Hosting/Utils/PasswordGenerator.cs#L11

File: tests/Aspire.Playground.Tests/Aspire.Playground.Tests.csproj (hunk 11 above)
[the remark this answers is not in the document]
Comment (eerhardt) [reply tests/Aspire.Playground.Tests/Aspire.Playground.Tests.csproj:28]: Ok, we can leave NodeJS for now since we reference a type in it.

File: tests/helix/send-to-helix-inner.proj (hunk 6 above)
Comment (radical) [reply tests/helix/send-to-helix-inner.proj:13]: hmm this can be extracted as a separate change. It is essentially making sure that we don't bundle, and set up things for running playwright tests when it is disabled for that platform - linux, in this case.

File: tests/Aspire.Playground.Tests/Aspire.Playground.Tests.csproj (hunk 9 above)
Comment (radical) [reply tests/Aspire.Playground.Tests/Aspire.Playground.Tests.csproj:25]: hmm I think this can be reworked so this project builds on the build machine. Then we can build the playground projects on helix, and point the tests to the bindir for these.

File: playground/AzureStorageEndToEnd/AzureStorageEndToEnd.AppHost/Program.cs [hunk 13]
```diff
@@ -13,12 +13,14 @@
        .WithExternalHttpEndpoints()
        .WithReference(blobs);
 
+#if !SKIP_DASHBOARD_REFERENCE
```
Comment (joperezr) [inline playground/AzureStorageEndToEnd/AzureStorageEndToEnd.AppHost/Program.cs:16]: NIT: Should the comment bellow be updated to say "pass /p:SkipDashboardProjectReference=true" instead of "Comment out this code"?

File: playground/mongo/Mongo.ApiService/Program.cs [hunk 14]
```diff
@@ -12,6 +12,7 @@
 
 var app = builder.Build();
 
+app.MapDefaultEndpoints();
```
Comment (joperezr) [inline playground/mongo/Mongo.ApiService/Program.cs:15]: It's actually interesting that we weren't calling this in the past but we were Adding Service Defaults. Do you happen to know why? Why do we need it now?

Comment (radical) [reply playground/mongo/Mongo.ApiService/Program.cs:15]: We need it to get the `/alive`, and `/health` endpoints being used by the tests. When running it manually, it wasn't useful I'm guessing.

File: tests/Aspire.Playground.Tests/Aspire.Playground.Tests.csproj (hunk 9 above)
Comment (radical) [reply tests/Aspire.Playground.Tests/Aspire.Playground.Tests.csproj:25]: This will need some reworking. I have added it as a follow up task in https://github.com/dotnet/aspire/issues/5210 . We can get the tests checked-in before that.

File: tests/Aspire.Playground.Tests/.runsettings (hunk 5 above)
Comment (radical) [reply tests/Aspire.Playground.Tests/.runsettings:1]: This will get covered by the follow up task of building on the build machine, and only running on helix.

File: playground/AzureStorageEndToEnd/AzureStorageEndToEnd.AppHost/Program.cs (hunk 13 above)
Comment (radical) [reply playground/AzureStorageEndToEnd/AzureStorageEndToEnd.AppHost/Program.cs:16]: done.

File: tests/helix/send-to-helix-inner.proj (hunk 6 above)
Comment (radical) [reply tests/helix/send-to-helix-inner.proj]: Reverted.

File: playground/Directory.Build.targets [hunk 15]
```diff
@@ -1,11 +1,14 @@
...
     <ProjectReference Include="$(RepoRoot)src\Aspire.Dashboard\Aspire.Dashboard.csproj" />
   </ItemGroup>
 
-  <!-- When $(BuildForTest)==true, importing Aspire.RepoTesting.targets is needed. But import
-       only when in-repo. For the out-of-repo case a parent Directory.Build.targets does the import -->
-  <Import Project="$(TestsSharedRepoTestingDir)Aspire.RepoTesting.targets" Condition="'$(RepoRoot)' != '' and '$(BuildForTest)' != 'true'" />
+  <!-- Import only when in-repo. For the out-of-repo case a parent Directory.Build.targets does the import -->
+  <Import Project="$(TestsSharedRepoTestingDir)Aspire.RepoTesting.targets" Condition="'$(RepoRoot)' != ''" />
+
+  <PropertyGroup Condition="'$(TestsRunningOutsideOfRepo)' == 'true' or '$(SkipDashboardProjectReference)' == 'true'">
```
Comment (eerhardt) [inline playground/Directory.Build.targets:11]:
```suggestion
  <PropertyGroup Condition="'$(SkipDashboardProjectReference)' == 'true'">
```

When are these tests running outside of the repo and SkipDashboardProjectReference is not set?

File: tests/Aspire.Playground.Tests/Aspire.Playground.Tests.csproj [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+
+    <TestArchiveTestsDir>$(TestArchiveTestsDirForBuildOnHelixTests)</TestArchiveTestsDir>
+
+    <PlaygroundSourceDir>$(MSBuildThisFileDirectory)..\..\playground\</PlaygroundSourceDir>
+    <TestsSharedDir>$(MSBuildThisFileDirectory)..\Shared\</TestsSharedDir>
+    <RunSettingsFilePath>$(MSBuildThisFileDirectory).runsettings</RunSettingsFilePath>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <Compile Condition="'$(RepoRoot)' != ''" Include="$(RepoRoot)src\Aspire.Hosting\Utils\PasswordGenerator.cs" />
```
Comment (eerhardt) [inline tests/Aspire.Playground.Tests/Aspire.Playground.Tests.csproj:23]: How does this build when `RepoRoot` isn't set? `WithRandomParameterValues` needs this code in order to build.

File: tests/Aspire.Playground.Tests/AppHostTests.cs [hunk 17]
```diff
@@ -0,0 +1,218 @@
...
+namespace Aspire.Playground.Tests;
+
+public class AppHostTests
+{
+    private readonly TestOutputWrapper _testOutput;
+    private static readonly string? s_appHostNameFilter = Environment.GetEnvironmentVariable("TEST_PLAYGROUND_APPHOST_FILTER");
+
+    public AppHostTests(ITestOutputHelper testOutput)
+    {
+        this._testOutput = new TestOutputWrapper(testOutput);
```
Comment (eerhardt) [inline tests/Aspire.Playground.Tests/AppHostTests.cs:23]: Why do we need this `TestOutputWrapper`? Why can't we use the `ITestOutputHelper` directly?

Comment (eerhardt) [inline tests/Aspire.Playground.Tests/AppHostTests.cs:23]: (nit)

```suggestion
        _testOutput = new TestOutputWrapper(testOutput);
```

**Review by eerhardt — APPROVED** [review]
This is great work, @radical. Nice job.

Let's get this in and get more tests enabled.

File: tests/Aspire.Playground.Tests/Aspire.Playground.Tests.csproj (hunk 16 above)
Comment (radical) [reply tests/Aspire.Playground.Tests/Aspire.Playground.Tests.csproj:23]: We copy the source to the test directory in the archive for helix, so it gets picked up by msbuild as part of collecting the all `.cs` files.

File: tests/Aspire.Playground.Tests/AppHostTests.cs (hunk 17 above)
Comment (radical) [reply tests/Aspire.Playground.Tests/AppHostTests.cs:23]: This is useful to see the output streaming to `testOutputHelper` when the test run might be taking long. But in the case of these tests most of the output is being logged to the console anyway, so we can drop this.

**Comment by radical** [comment]
Merging this, and will open follow up PRs for the suggestions.
